Shelfice autodiff

timothyas opened the pull request:
## What changes does this PR introduce?
(Bug fix, feature, docs update, ...)

Implements shelfice+autodiff related cleanup discussed in #349 and #440:
- Get rid of `xx_shifwflx` and the CPP macro `ALLOW_SHIFWFLX_CONTROL`. This is since replaced with gentim2d, and the only substantial change was in `shelfice_thermodynamics`, where `xx_shifwflx_loc` (a local variable) is assigned `xx_gentim2d`. This is already done, just had to remove the code block that would read in `xx_shifwflx_file`, etc ... Note that with this gone, the routine `shelfice_cost_shifwflx` is no longer necessary since this just computes the regularization term for this control variable. So I deleted it. This is done within `ctrl_cost_driver` and `ctrl_cost_gen2d` for all generic controls.
- remove the `#ifndef ALLOW_AUTODIFF` macro that was oddly surrounding the load anomaly update at the end of shelfice thermodynamics, and in fact I moved this block of code to `shelfice_step_icemass`. This seems correct, but I'd like someone to check me on that.
- remove the `STOP` statement in `shelfice_forcing_surf` related to autodiff as discussed in #349 . Now, the validity of the adjoint model in the case of using the real freshwater flux with the iceshelf remains to be studied. However, this is a bit more involved and better evaluated on a case-by-case basis. I have found that the adjoint runs as expected, so I think the stop is unnecessary.

## What is the current behaviour? 
(You can also link to an open issue here)

#349 #440

## What is the new behaviour 
(if this is a feature change)?

- gentim2d in favor of individual control variables, control package type code removed from `shelfice_thermodynamics`
- a probably correct load anomaly for people using autodiff and iceshelf mass stepping
- realfwf + shelfice + adjoints

Closes #349 

## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)


## Other information:


## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

Not sure if all of these are necessary in the tag-index but here are suggestions:

- Remove ALLOW_SHIFWFLX_CONTROL and all xx_shifwflx related code in favor of gentim2d, which is already implemented
- Remove unnecessary STOP in shelfice_forcing_surf when realFreshWaterFlux=.TRUE. and ALLOW_AUTODIFF is defined
- Remove #ifndef ALLOW_AUTODIFF around shelfice load anomaly update in shelfice_thermodynamics

timothyas commented:
Slapping the draft label on this since it's not passing the isomip AD test and I've run out of time to work on it for today. Be back on it asap.

mjlosch commented:
@timothyas there are many (20) extra recomputations that can be fixed by adding 
```
CADJ STORE shelficeFreshWaterFlux = tapelevX, key = ilev_X
CADJ STORE shelficeLoadAnomaly    = tapelevX, key = ilev_X
```
to `shelfice_ad_check_levX_dir.h` (`X=2,3,4`). This also makes isomip AD pass. 
I am not sure why this is suddenly necessary. Do you have any idea?

mjlosch commented on pkg/shelfice/shelfice_step_icemass.F:
```diff
@@ -100,6 +100,18 @@ SUBROUTINE SHELFICE_STEP_ICEMASS(
 
        _EXCH_XY_RL( shelficeMass, myThid )
 
+C--  Calculate new loading anomaly (in case the ice-shelf mass was updated)
```
I checked that this does no cause the recomputations, but if we want to update `shelficeLoadAnomaly` in each timestep (i.e. it becomes active in a TAF-sense), then it would probably make sense to move the corresponding initialisation code starting here https://github.com/MITgcm/MITgcm/blob/0f3d23268a0f2aa734ede108df9096f60b631b2a/pkg/shelfice/shelfice_init_fixed.F#L135 to shelfice_init_varia.F, right?

mjlosch replied:
OK, I take that back, because then we will have an issue with initialising `phi0surf`.

mjlosch commented on pkg/shelfice/shelfice_thermodynamics.F:
```diff
@@ -122,12 +122,9 @@ SUBROUTINE SHELFICE_THERMODYNAMICS(
 #ifdef ALLOW_GENTIM2D_CONTROL
       INTEGER iarr
 #endif
```
Maybe you can merge this into one ifdef-block? Further down (https://github.com/MITgcm/MITgcm/blob/0f3d23268a0f2aa734ede108df9096f60b631b2a/pkg/shelfice/shelfice_thermodynamics.F#L167), we could probably remove the second re-initialization of shiTransCoeffT/S=0, but it's not tested anywhere.

timothyas replied:
Just to follow up on this, I tested removing the second re-initialization of shiTransCoeffT/S in my own setup and (as expected) there's no difference with it or without it.

mjlosch replied:
@timothyas I think that this test is enough and we can resolve this comment.

mjlosch commented:
@timothyas Sorry for the chaotic review. I think it is probably the right to move the update `shelficeLoadAnomaly` to shelfice_step_icemass, as it is only relevant if ice mass changes.

mjlosch commented:
> there are many (20) extra recomputations

As I already have the fix, would you like me to push it, or would you prefer to first test this yourself?

timothyas commented:
Thanks @mjlosch! Sorry for the delay. Yes, please go ahead and push - thanks for digging into this

timothyas commented:
Hey all, thanks for your patience on this PR. I've finally revived this, and I think it's ready for review. To answer your earlier question @mjlosch: 

> @timothyas there are many (20) extra recomputations that can be fixed by adding
> 
> ```
> CADJ STORE shelficeFreshWaterFlux = tapelevX, key = ilev_X
> CADJ STORE shelficeLoadAnomaly    = tapelevX, key = ilev_X
> ```
> 
> to `shelfice_ad_check_levX_dir.h` (`X=2,3,4`). This also makes isomip AD pass. I am not sure why this is suddenly necessary. Do you have any idea?

I have done some local tests to confirm that those warnings are now showing up for two reasons. First, because I have removed the `STOP` statement in shelfice_forcing_surf (discussed in #349) https://github.com/MITgcm/MITgcm/blob/2459b396bd513b72c430a977c2d62d020b608a82/pkg/shelfice/shelfice_forcing_surf.F#L103-L105

 And secondly because I have removed the strange `#ifndef ALLOW_AUTODIFF` statement surrounding the pressure load anomaly update in shelfice_thermodynamics, which in this PR is moved to shelfice_step_icemass: https://github.com/MITgcm/MITgcm/blob/2459b396bd513b72c430a977c2d62d020b608a82/pkg/shelfice/shelfice_thermodynamics.F#L731-L744

Since both of these lines of code were not present when autodiff was being used, we never saw those warnings. Now I think we're in good shape though.

timothyas commented:
Thanks for tidying up the `STORE` statements @mjlosch!

mjlosch commented on pkg/shelfice/shelfice_cost_final.F:
```diff
@@ -67,12 +63,7 @@ SUBROUTINE SHELFICE_COST_FINAL( myThid )
-      CALL SHELFICE_COST_SHIFWFLX( myThid )
-#endif /* ALLOW_SHIFWFLX_CONTROL and ALLOW_SHIFWFLX_COST_CONTRIBUTION */
-#else
+#ifndef ALLOW_ECCO
```
would it make sense to replace the CPP options by an `IF (.NOT. useECCO ) THEN` block?

mjlosch replied:
PS. I can push such a change if you think it makes sense.

timothyas replied:
That sounds fine to me. I went with the CPP since that is what was there, but I think what you're proposing makes sense (using runtime flags and including `useShelfice`). 

Push away - thanks!

mjlosch commented:
@timothyas this looks great! So many simplifications. (I am look so much forward to getting rid of the ECCO_CTRL_DEPRECATED code.)

timothyas commented:
Thanks @mjlosch! I totally agree :)

mjlosch commented on pkg/shelfice/shelfice_step_icemass.F:
```diff
@@ -100,6 +100,18 @@ SUBROUTINE SHELFICE_STEP_ICEMASS(
        _EXCH_XY_RL( shelficeMass, myThid )
 
+C--  Calculate new loading anomaly (in case the ice-shelf mass was updated)
+       DO bj = myByLo(myThid), myByHi(myThid)
+        DO bi = myBxLo(myThid), myBxHi(myThid)
+         DO j = 1-OLy, sNy+OLy
+          DO i = 1-OLx, sNx+OLx
+           shelficeLoadAnomaly(i,j,bi,bj) = gravity
+     &      *( shelficeMass(i,j,bi,bj) + rhoConst*Ro_surf(i,j,bi,bj) )
+          ENDDO
+         ENDDO
+        ENDDO
+       ENDDO
```
It looks like this addition causes the OAD shelfice test to fail at the compilation step. As a quick fix, you could enclose it with ifndef ALLOW_OPENAD like this
```suggestion
#ifndef ALLOW_OPENAD
       DO bj = myByLo(myThid), myByHi(myThid)
        DO bi = myBxLo(myThid), myBxHi(myThid)
         DO j = 1-OLy, sNy+OLy
          DO i = 1-OLx, sNx+OLx
           shelficeLoadAnomaly(i,j,bi,bj) = gravity
     &      *( shelficeMass(i,j,bi,bj) + rhoConst*Ro_surf(i,j,bi,bj) )
          ENDDO
         ENDDO
        ENDDO
       ENDDO
#endif       
```

timothyas replied:
Thanks @mjlosch. I went this route in [5adf770](https://github.com/MITgcm/MITgcm/pull/495/commits/5adf770abcba6ab9f374cef63094e6301937d438) and lo and behold the tests are passing.

Now moving forward, I think the minimum is to add a warning or error (my preference on the latter) to the user if the compile time option `#define ALLOW_OPENAD` and runtime option `SHELFICEMassStepping` are true. It is annoying to put band-aids over the code like this. But, this brute force fix would not change the behavior for OpenAD users within this PR. If an OpenAD user would like this to work, maybe that could get pursued in a future PR, what do others (e.g. @jm-c, @dngoldberg) think about that?

jm-c replied:
I am fine with this specific change (#ifndef ALLOW_OPENAD block) but I agree that it would be nice to add a STOP + 1 or 2 error message line in shelfice_check.F

mjlosch commented on pkg/shelfice/shelfice_thermodynamics.F:
```diff
@@ -727,22 +697,6 @@ SUBROUTINE SHELFICE_THERMODYNAMICS(
        CALL SHELFICE_STEP_ICEMASS( myTime, myIter, myThid )
       ENDIF
 
-C--  Calculate new loading anomaly (in case the ice-shelf mass was updated)
-#ifndef ALLOW_AUTODIFF
-c     IF ( SHELFICEloadAnomalyFile .EQ. ' ' ) THEN
-       DO bj = myByLo(myThid), myByHi(myThid)
-        DO bi = myBxLo(myThid), myBxHi(myThid)
-         DO j = 1-OLy, sNy+OLy
-          DO i = 1-OLx, sNx+OLx
-           shelficeLoadAnomaly(i,j,bi,bj) = gravity
-     &      *( shelficeMass(i,j,bi,bj) + rhoConst*Ro_surf(i,j,bi,bj) )
-          ENDDO
-         ENDDO
-        ENDDO
-       ENDDO
-c     ENDIF
-#endif /* ndef ALLOW_AUTODIFF */
-
```
This is now moved to shelfice_step_icemass.F where it causes said problems with OpenAD, but the real reason is the ifndef ALLOW_AUTODIFF that was not moved along with this block. I.e., this probably never worked with OpenAD, but we didn't notice.

timothyas replied:
Thanks @mjlosch

jm-c commented on pkg/shelfice/shelfice_thermodynamics.F:
```diff
@@ -727,22 +697,6 @@ SUBROUTINE SHELFICE_THERMODYNAMICS(
-C--  Calculate new loading anomaly (in case the ice-shelf mass was updated)
-#ifndef ALLOW_AUTODIFF
-c     IF ( SHELFICEloadAnomalyFile .EQ. ' ' ) THEN
-       DO bj = myByLo(myThid), myByHi(myThid)
```
Just a comment here: This block (to update shelficeLoadAnomaly) should have been put within "IF SHELFICEMassStepping / ENDIF" when new run-time flag "SHELFICEMassStepping" was introduced (in Dec 2014).
This is done here by moving this block to S/R SHELFICE_STEP_ICEMASS which is only called when SHELFICEMassStepping=T

jm-c commented on model/src/the_main_loop.F:
```diff
@@ -298,20 +298,21 @@ SUBROUTINE THE_MAIN_LOOP( myTime, myIter, myThid )
-CADJ STORE shelficeHeatFlux       = tapelev_init, key=1
-CADJ STORE shiTransCoeffT         = tapelev_init, key=1
-CADJ STORE shiTransCoeffS         = tapelev_init, key=1
+# ifdef ALLOW_SHELFICE
```
The condition for adding these store dirs has changed from ALLOW_SHELFICE & ALLOW_GENARR2D_CONTROL
to just simply ALLOW_SHELFICE. It's not very clear to me how this is related to the other set of changes in this PR.

mjlosch replied:
This change is just cleaning up the cpp-flags a little. There is no reason to have these stores depend on `ALLOW_GENARR2D_CONTROL`, because
1. even though these are 2D fields, some of these fields depend on code withing `pkg/shelfice`, so it is not clear a-priori if these stores are necessary or not.
2. taf will most likely figure this out by itself, if these stores are necessary or not

Some similar reasoning applies to the STREAMICE change below.

jm-c replied:
@mjlosch Thanks, this is useful.

jm-c commented on model/src/the_main_loop.F:
```diff
@@ -298,20 +298,21 @@ SUBROUTINE THE_MAIN_LOOP( myTime, myIter, myThid )
+CADJ STORE shiTransCoeffS         = tapelev_init, key=1, kind=isbyte
 # endif
-# if (defined ALLOW_STREAMICE) && (defined ALLOW_GENARR2D_CONTROL)
+# ifdef ALLOW_STREAMICE
```
And similarly, not clear why these pkg/streamice changes are related to the other changes in this PR.

jm-c reviewed (approved):
This looks good and even improve the forward code (independent of autodiff). I left 2 comments/questions regarding storage dir in the_main_loop.F, probably nothing to change there but just curious why.

jm-c commented:
I think this is ready to be merged in. I could do this tomorrow if nobody disagree.
